db/database_collection.go

```diff
@@ -0,0 +1,167 @@
+}
+
+// DatabaseCollectionWithUser represents CouchDB database. A new instance is created for each request,
+// so this struct does not have to be thread-safe.
+type DatabaseCollectionWithUser struct {
+	*DatabaseCollection
+	user auth.User
+}
+
+// AllowConflicts allows different revisions of a single document to be pushed. This is controlled at the database level.
+func (c *DatabaseCollection) AllowConflicts() bool {
+	if c.dbCtx.Options.AllowConflicts != nil {
+		return *c.dbCtx.Options.AllowConflicts
+	}
+	return base.DefaultAllowConflicts
+}
+
+// Authenticator returns authentication options associated with the collection's database.
+func (c *DatabaseCollection) Authenticator(ctx context.Context) *auth.Authenticator {
+	return c.dbCtx.Authenticator(ctx)
+}
+
+// backupOldRev creates a temporary copy of a old revision body when resolving conflicts. This is controlled at the database level.
+func (c *DatabaseCollection) backupOldRev() bool {
+	return c.dbCtx.Options.ImportOptions.BackupOldRev
+
+}
+
+// changeCache returns the set of channels in memory. This is currently at the database level.
+func (c *DatabaseCollection) changeCache() *changeCache {
+	return c.dbCtx.changeCache
+}
+
+// channelMapper runs the javascript sync function. The is currently at the database level.
+func (c *DatabaseCollection) channelMapper() *channels.ChannelMapper {
+	return c.dbCtx.ChannelMapper
+}
+
+// DbStats are stats that correspond to database level collections.
+func (c *DatabaseCollection) dbStats() *base.DbStats {
+	return c.dbCtx.DbStats
+}
+
+// deltaSyncEnabled returns true if delta sync is enabled. This is controlled at the database level.
+func (c *DatabaseCollection) deltaSyncEnabled() bool {
+	return c.dbCtx.Options.DeltaSyncOptions.Enabled
+}
+
+// deltaSyncRevMaxAgeSeconds is the number of seconds that old revisions will be in memory. This is controlled at database level.
+func (c *DatabaseCollection) deltaSyncRevMaxAgeSeconds() uint32 {
+	return c.dbCtx.Options.DeltaSyncOptions.RevMaxAgeSeconds
+}
+
+// eventMgr manages nofication events. This is controlled at database level.
+func (c *DatabaseCollection) eventMgr() *EventManager {
+	return c.dbCtx.EventMgr
+}
+
+// FlushRevisionCacheForTest creates a new revision cache. This is currently at the database level. Only use this in test code.
+func (c *DatabaseCollection) FlushRevisionCacheForTest() {
+	c.revisionCache = NewRevisionCache(
+		c.dbCtx.Options.RevisionCacheOptions,
+		c,
+		c.dbStats().Cache(),
+	)
+
+}
+
+// ForceAPIForbiddenErrors returns true if we return 403 vs empty docs. This is controlled at the database level.
+func (c *DatabaseCollection) ForceAPIForbiddenErrors() bool {
+	return c.dbCtx.Options.UnsupportedOptions != nil && c.dbCtx.Options.UnsupportedOptions.ForceAPIForbiddenErrors
+}
+
+// importFilter returns the sync function.
+// FIXME: collections - this needs to be moved to per-collection
```

Comment:
For all these FIXME, would be preferable to include a link to the associated CBG ticket (and create if needed).